In [1]:
#Import data using pydataset
from pydataset import data as d

In [2]:
d('mpg', show_doc=True) # view the documentation for the dataset
mpg = d('mpg') # load the dataset and store it in a variable

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




#  <center>1</center></h1>
<center>Copy the code from the lesson to create a dataframe full of student grades.</center></h1>

In [3]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

In [4]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     12 non-null     object
 1   math     12 non-null     int64 
 2   english  12 non-null     int64 
 3   reading  12 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 512.0+ bytes


In [6]:
# 1.a Create a column named passing_english that indicates whether each student has a passing grade in english.
df['passing_english'] = df.math > 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80            False
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
9   Richard    69       80       94            False
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

In [9]:
# 1.b Sort the english grades by the passing_english column. 
'''How are duplicates handled? 
Duplicates are handled by alphabetical order of the first column of the record'''

df1b = df.sort_values(by='passing_english')
df1b

name  math  english  reading  passing_english
0     Sally    62       85       80            False
9   Richard    69       80       94            False
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81             True
7     Marie    93       63       90             True
8    Albert    92       62       87             True
10    Isaac    92       99       93             True
11     Alan    92       62       72             True

In [11]:
# 1.c Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)
df1c = df1b.sort_values(by='name')
df1c

name  math  english  reading  passing_english
4       Ada    77       92       98             True
11     Alan    92       62       72             True
8    Albert    92       62       87             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
7     Marie    93       63       90             True
9   Richard    69       80       94            False
0     Sally    62       85       80            False
2     Suzie    94       74       95             True
6    Thomas    82       64       81             True

In [14]:
# 1.d Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.
df1d = df.sort_values(by=['passing_english', 'english'])
df1d

name  math  english  reading  passing_english
9   Richard    69       80       94            False
0     Sally    62       85       80            False
8    Albert    92       62       87             True
11     Alan    92       62       72             True
7     Marie    93       63       90             True
6    Thomas    82       64       81             True
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

In [17]:
# 1.e Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.
df['OVERALL'] = ((df.math + df.english + df.reading) / 3).round(decimals=2)
df['Passed'] = df.OVERALL > 70
df

name  math  english  reading  passing_english  OVERALL  Passed
0     Sally    62       85       80            False    75.67    True
1      Jane    88       79       67             True    78.00    True
2     Suzie    94       74       95             True    87.67    True
3     Billy    98       96       88             True    94.00    True
4       Ada    77       92       98             True    89.00    True
5      John    79       76       93             True    82.67    True
6    Thomas    82       64       81             True    75.67    True
7     Marie    93       63       90             True    82.00    True
8    Albert    92       62       87             True    80.33    True
9   Richard    69       80       94            False    81.00    True
10    Isaac    92       99       93             True    94.67    True
11     Alan    92       62       72             True    75.33    True

## 

# <center>2</center></h1>

<center>Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:</center></h1>

#### How many rows and columns are there?
There are 234 rows and 11 colums

In [24]:
mpg.describe

<bound method NDFrame.describe of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    

#### What are the data types of each column? 

In [28]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

#### Summarize the dataframe with .info and .describe

In [29]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

In [30]:
mpg.describe

<bound method NDFrame.describe of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    

In [35]:
# Rename the cty column to city.
mpg = mpg.rename(columns={'cty':'city'})

In [37]:
# Rename the hwy column to highway.
mpg = mpg.rename(columns={'hwy':'highway'})

In [58]:
# Do any cars have better city mileage than highway mileage?
[mpg.city > mpg.highway]
# mpg[mpg['city'] > mpg['highway']]
    
    

[1      False
 2      False
 3      False
 4      False
 5      False
        ...  
 230    False
 231    False
 232    False
 233    False
 234    False
 Length: 234, dtype: bool]

In [ ]:
mpg_higher_city = mpg[mpg.city > mpg.highway]

In [59]:
mpg_higher_city

NameError: name 'mpg_higher_city' is not defined

In [55]:
# Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.
mpg['milage_difference'] = mpg.highway - city

NameError: name 'city' is not defined

In [ ]:
# Which car (or cars) has the highest mileage difference?


In [ ]:
# Which compact class car has the lowest highway mileage? The best?

In [ ]:
# Create a column named average_mileage that is the mean of the city and highway mileage.


In [ ]:
# Which dodge car has the best average mileage? The worst?

# <center>3</center></h1>

<center>Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:</center></h1>

In [ ]:
from pydataset import data 


In [ ]:
# How many rows and columns are there?


In [ ]:
# What are the data types?


In [ ]:
# Summarize the dataframe with .info and .describe


In [ ]:
# What is the the weight of the fastest animal?


In [54]:
# What is the overal percentage of specials?


In [ ]:
# How many animals are hoppers that are above the median speed? What percentage is this?
